src/precice/impl/SolverInterfaceImpl.cpp

```diff
@@ -976,33 +976,37 @@ void SolverInterfaceImpl::mapWriteDataFrom(
 {
   PRECICE_TRACE(fromMeshID);
   PRECICE_VALIDATE_MESH_ID(fromMeshID);
-  impl::MeshContext &   context        = _accessor->meshContext(fromMeshID);
-  impl::MappingContext &mappingContext = context.fromMappingContext;
-  if (not mappingContext.mapping) {
+  impl::MeshContext &context = _accessor->meshContext(fromMeshID);
+
+  if (context.fromMappingContexts.empty()) {
     PRECICE_ERROR("You attempt to \"mapWriteDataFrom\" mesh "
                   << context.mesh->getName()
                   << ", but there is no mapping from this mesh configured."
                      "Maybe you don't want to call this function at all or you forgot to configure the mapping.");
     return;
   }
+
   double time = _couplingScheme->getTime();
   performDataActions({action::Action::WRITE_MAPPING_PRIOR}, time, 0, 0, 0);
-  if (not mappingContext.mapping->hasComputedMapping()) {
-    PRECICE_DEBUG("Compute mapping from mesh \"" << context.mesh->getName() << "\"");
-    mappingContext.mapping->computeMapping();
-  }
-  for (impl::DataContext &context : _accessor->writeDataContexts()) {
-    if (context.mesh->getID() == fromMeshID) {
-      int inDataID             = context.fromData->getID();
-      int outDataID            = context.toData->getID();
-      context.toData->values() = Eigen::VectorXd::Zero(context.toData->values().size());
-      PRECICE_DEBUG("Map data \"" << context.fromData->getName()
-                                  << "\" from mesh \"" << context.mesh->getName() << "\"");
-      PRECICE_ASSERT(mappingContext.mapping == context.mappingContext.mapping);
-      mappingContext.mapping->map(inDataID, outDataID);
+
+  for (impl::MappingContext &mappingContext : context.fromMappingContexts) {
+    if (not mappingContext.mapping->hasComputedMapping()) {
+      PRECICE_DEBUG("Compute mapping from mesh \"" << context.mesh->getName() << "\"");
+      mappingContext.mapping->computeMapping();
     }
+    for (impl::DataContext &context : _accessor->writeDataContexts()) {
+      if (context.mesh->getID() == fromMeshID) {
```

Comment:
We could use a guard here to prevent nesting.
```suggestion
      if (context.mesh->getID() != fromMeshID) {
          continue;
      }
```